File: android/core/src/main/kotlin/com/iblocker/core/dns/DnsMessage.kt
package com.iblocker.core.dns

/** Well-known DNS record types the app cares about (display + blocking policy). */
object DnsRecordType {
    const val A = 1
    const val NS = 2
    const val CNAME = 5
    const val SOA = 6
    const val PTR = 12
    const val MX = 15
    const val TXT = 16
    const val AAAA = 28
    const val SRV = 33
    const val OPT = 41
    const val SVCB = 64
    const val HTTPS = 65

    fun name(qtype: Int): String = when (qtype) {
        A -> "A"
        NS -> "NS"
        CNAME -> "CNAME"
        SOA -> "SOA"
        PTR -> "PTR"
        MX -> "MX"
        TXT -> "TXT"
        AAAA -> "AAAA"
        SRV -> "SRV"
        OPT -> "OPT"
        SVCB -> "SVCB"
        HTTPS -> "HTTPS"
        else -> "TYPE$qtype"
    }
}

/** A parsed DNS query (QR=0, one question). */
class DnsQuery(
    val transactionID: Int,
    val flags: Int,
    /** Lowercased dotted question name, no trailing dot. */
    val questionName: String,
    /**
     * Byte-exact question section (name + qtype + qclass) as it appeared on
     * the wire, preserving the original case (DNS 0x20 randomization).
     */
    val rawQuestion: ByteArray,
    val qtype: Int,
    val qclass: Int,
    /** True when the query carried anything in the additional section (in practice: EDNS0 OPT). */
    val hasAdditionalRecords: Boolean,
    /** The complete original message, for untouched upstream forwarding. */
    val raw: ByteArray,
) {
    val recursionDesired: Boolean get() = flags and 0x0100 != 0
}

object DnsParser {

    /**
     * Parses a standard query with exactly one question.
     * Throws for responses and multi-question messages; callers should
     * forward those upstream untouched rather than fail the lookup.
     */
    @Throws(DnsWireException::class)
    fun parseQuery(data: ByteArray): DnsQuery {
        if (data.size < 12) throw DnsWireException(DnsWireError.TRUNCATED)

        val id = u16(data, 0)
        val flags = u16(data, 2)
        if (flags and 0x8000 != 0) throw DnsWireException(DnsWireError.NOT_A_QUERY)

        val qdcount = u16(data, 4)
        val arcount = u16(data, 10)
        if (qdcount != 1) throw DnsWireException(DnsWireError.UNSUPPORTED_QUESTION_COUNT)

        val (name, nameLength) = DnsNameCodec.decode(data, 12)
        val qtypeOffset = 12 + nameLength
        if (qtypeOffset + 4 > data.size) throw DnsWireException(DnsWireError.TRUNCATED)

        return DnsQuery(
            transactionID = id,
            flags = flags,
            questionName = name,
            rawQuestion = data.copyOfRange(12, qtypeOffset + 4),
            qtype = u16(data, qtypeOffset),
            qclass = u16(data, qtypeOffset + 2),
            hasAdditionalRecords = arcount > 0,
            raw = data,
        )
    }

    private fun u16(bytes: ByteArray, offset: Int): Int =
        ((bytes[offset].toInt() and 0xFF) shl 8) or (bytes[offset + 1].toInt() and 0xFF)
}

File: ios/IBlockerKit/Sources/IBlockerKit/DNS/DNSMessage.swift
import Foundation

/// Well-known DNS record types the app cares about (display + blocking policy).
public enum DNSRecordType {
    public static let a: UInt16 = 1
    public static let ns: UInt16 = 2
    public static let cname: UInt16 = 5
    public static let soa: UInt16 = 6
    public static let ptr: UInt16 = 12
    public static let mx: UInt16 = 15
    public static let txt: UInt16 = 16
    public static let aaaa: UInt16 = 28
    public static let srv: UInt16 = 33
    public static let opt: UInt16 = 41
    public static let svcb: UInt16 = 64
    public static let https: UInt16 = 65

    public static func name(_ qtype: UInt16) -> String {
        switch qtype {
        case a: return "A"
        case ns: return "NS"
        case cname: return "CNAME"
        case soa: return "SOA"
        case ptr: return "PTR"
        case mx: return "MX"
        case txt: return "TXT"
        case aaaa: return "AAAA"
        case srv: return "SRV"
        case opt: return "OPT"
        case svcb: return "SVCB"
        case https: return "HTTPS"
        default: return "TYPE\(qtype)"
        }
    }
}

/// A parsed DNS query (QR=0, one question).
public struct DNSQuery: Sendable {
    public let transactionID: UInt16
    public let flags: UInt16
    /// Lowercased dotted question name, no trailing dot.
    public let questionName: String
    /// Byte-exact question section (name + qtype + qclass) as it appeared on
    /// the wire, preserving the original case (DNS 0x20 randomization).
    public let rawQuestion: [UInt8]
    public let qtype: UInt16
    public let qclass: UInt16
    /// True when the query carried anything in the additional section
    /// (in practice: an EDNS0 OPT record).
    public let hasAdditionalRecords: Bool
    /// The complete original message, for untouched upstream forwarding.
    public let raw: Data

    public var recursionDesired: Bool { flags & 0x0100 != 0 }
}

public enum DNSParser {

    /// Parses a standard query with exactly one question.
    /// Throws for responses and multi-question messages; callers should
    /// forward those upstream untouched rather than fail the lookup.
    public static func parseQuery(_ data: Data) throws -> DNSQuery {
        let bytes = [UInt8](data)
        guard bytes.count >= 12 else { throw DNSWireError.truncated }

        let id = UInt16(bytes[0]) << 8 | UInt16(bytes[1])
        let flags = UInt16(bytes[2]) << 8 | UInt16(bytes[3])
        guard flags & 0x8000 == 0 else { throw DNSWireError.notAQuery }

        let qdcount = Int(bytes[4]) << 8 | Int(bytes[5])
        let arcount = Int(bytes[10]) << 8 | Int(bytes[11])
        guard qdcount == 1 else { throw DNSWireError.unsupportedQuestionCount }

        let (name, nameLength) = try DNSNameCodec.decode(bytes, at: 12)
        let qtypeOffset = 12 + nameLength
        guard qtypeOffset + 4 <= bytes.count else { throw DNSWireError.truncated }

        let qtype = UInt16(bytes[qtypeOffset]) << 8 | UInt16(bytes[qtypeOffset + 1])
        let qclass = UInt16(bytes[qtypeOffset + 2]) << 8 | UInt16(bytes[qtypeOffset + 3])

        return DNSQuery(
            transactionID: id,
            flags: flags,
            questionName: name,
            rawQuestion: Array(bytes[12..<(qtypeOffset + 4)]),
            qtype: qtype,
            qclass: qclass,
            hasAdditionalRecords: arcount > 0,
            raw: data
        )
    }
}
